### PR TITLE
[23.05] Backport mesh corruption fix

### PR DESCRIPTION
**Pull request by lukasz1992**
- Backport some mac80211 fixes

### CODE REVIEW

**Comment by hauke** [comment]
Please rebase this on top of #15827. I plan to merge #15827 soon and it probably contains multiple of your mac80211 fixes already. 
Pleas remove the merge commit from this PR.

File: package/kernel/mac80211/patches/subsys/332-wifi-mac80211-check-clear-fast-rx-for-non-4addr-sta-.patch [hunk 1]
```diff
@@ -0,0 +1,35 @@
+From: Felix Fietkau <nbd@nbd.name>
+Date: Sat, 16 Mar 2024 08:37:21 +0100
+Subject: [PATCH] wifi: mac80211: check/clear fast rx for non-4addr sta VLAN
```
Comment (hauke) [inline package/kernel/mac80211/patches/subsys/332-wifi-mac80211-check-clear-fast-rx-for-non-4addr-sta-.patch:3]: This patch was added in #15827

File: package/kernel/mac80211/patches/subsys/331-wifi-mac80211-only-call-drv_sta_rc_update-for-upload.patch [hunk 2]
```diff
@@ -0,0 +1,25 @@
+From: Felix Fietkau <nbd@nbd.name>
+Date: Wed, 21 Feb 2024 14:41:40 +0100
+Subject: [PATCH] wifi: mac80211: only call drv_sta_rc_update for uploaded
+ stations
```
Comment (hauke) [inline package/kernel/mac80211/patches/subsys/331-wifi-mac80211-only-call-drv_sta_rc_update-for-upload.patch:4]: This patch was added in #15827

File: package/kernel/mac80211/patches/subsys/337-mac80211-fix-sband-iftype-data-lookup-for-AP_VLAN.patch [hunk 3]
```diff
@@ -0,0 +1,23 @@
+From: Felix Fietkau <nbd@nbd.name>
+Date: Thu, 22 Jun 2023 18:02:25 +0200
+Subject: [PATCH] mac80211: fix sband iftype data lookup for AP_VLAN
```
Comment (hauke) [inline package/kernel/mac80211/patches/subsys/337-mac80211-fix-sband-iftype-data-lookup-for-AP_VLAN.patch:3]: This patch was added in #15827

**Comment by rany2** [comment]
P.S. when you cherry-pick, use `git cherry-pick -x` as it would add a nice `(cherry picked from commit <hash>)` to your message. I think it's a requirement for all backport commits.

**Comment by hauke** [comment]
> Edit: actually not sure if #15898 supersedes this PR, maybe @hauke should comment?

I took some of the commits added here which I think should get backported into #15898

**Comment by lukasz1992** [comment]
@hauke updated PR; excluded patches present here: https://github.com/openwrt/openwrt/pull/15898

File: target/linux/mediatek/patches-5.15/965-v6.10-net-ethernet-mtk_eth_soc-ppe-add-support-for-multipl.patch
Comment (rany2) [inline target/linux/mediatek/patches-5.15/965-v6.10-net-ethernet-mtk_eth_soc-ppe-add-support-for-multipl.patch:1]: You need to backport https://github.com/openwrt/openwrt/commit/8ef4d785829241e3a6f95c896de210fd5d3f073b as well if you want to add the `"generic: 6.6: (re-)add support multiple PPE to mtk_eth_soc"` patch, but personally I'd avoid that patch entirely for now given that it broke previously.

**Review by rany2 — CHANGES_REQUESTED** [review]

File: target/linux/mediatek/patches-5.15/965-v6.10-net-ethernet-mtk_eth_soc-ppe-add-support-for-multipl.patch
Comment (lukasz1992) [reply target/linux/mediatek/patches-5.15/965-v6.10-net-ethernet-mtk_eth_soc-ppe-add-support-for-multipl.patch:1]: There is no "break;" in mtk_eth_soc.c code in 5.15 version, patch also does not apply.
It looks like the regression appeared later than 5.15 with 23.05-SNAPSHOT patches.

@hauke what is your opinion on adding support for multiple PPE - keep it or remove?

**Comment by lukasz1992** [comment]
> P.S. when you cherry-pick, use `git cherry-pick -x` as it would add a nice `(cherry picked from commit <hash>)` to your message. I think it's a requirement for all backport commits.
> 
> Edit: actually not sure if #15898 supersedes this PR, maybe @hauke should comment?

I will try to update it today evening, thanks

**Comment by rmilecki** [comment]
After rebasing, your commit 0b92ea21a5d0 ("`hostapd: don't ignore probe-requests with invalid DSSS params`") seems to be just adding some unneeded lines. You can probably drop it.

File: target/linux/mediatek/patches-5.15/965-v6.10-net-ethernet-mtk_eth_soc-ppe-add-support-for-multipl.patch
Comment (rany2) [inline target/linux/mediatek/patches-5.15/965-v6.10-net-ethernet-mtk_eth_soc-ppe-add-support-for-multipl.patch:1]: > There is no "break;" in mtk_eth_soc.c code in 5.15 version, patch also does not apply.

There is, it's at line 159 in your patch

Comment (lukasz1992) [reply target/linux/mediatek/patches-5.15/965-v6.10-net-ethernet-mtk_eth_soc-ppe-add-support-for-multipl.patch:1]: You are right, thanks! I will fix it today

**Comment by lukasz1992** [comment]
> After rebasing, your commit [0b92ea2](https://github.com/openwrt/openwrt/commit/0b92ea21a5d001155e26f808b06765dbba77d7cc) ("`hostapd: don't ignore probe-requests with invalid DSSS params`") seems to be just adding some unneeded lines. You can probably drop it.

I do not know what happened on github, but if you preview the file - it is actually complete:
https://github.com/openwrt/openwrt/blob/0b92ea21a5d001155e26f808b06765dbba77d7cc/package/network/services/hostapd/patches/762-AP-don-t-ignore-probe-requests-with-invalid-DSSS-par.patch

**Comment by rany2** [comment]
> > After rebasing, your commit [0b92ea2](https://github.com/openwrt/openwrt/commit/0b92ea21a5d001155e26f808b06765dbba77d7cc) ("`hostapd: don't ignore probe-requests with invalid DSSS params`") seems to be just adding some unneeded lines. You can probably drop it.
> 
> I do not know what happened on github, but if you preview the file - it is actually complete: https://github.com/openwrt/openwrt/blob/0b92ea21a5d001155e26f808b06765dbba77d7cc/package/network/services/hostapd/patches/762-AP-don-t-ignore-probe-requests-with-invalid-DSSS-par.patch

GitHub is right, the patch is already included in the 23.05 branch: https://github.com/openwrt/openwrt/blob/openwrt-23.05/package/network/services/hostapd/patches/762-AP-don-t-ignore-probe-requests-with-invalid-DSSS-par.patch

File: package/kernel/mac80211/patches/subsys/317-mac80211-fix-mesh-id-corruption-on-32-bit-systems.patch [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+Content-Transfer-Encoding: 8bit
+
+Since the changed field size was increased to u64, mesh_bss_info_changed
+pulls invalid bits from the first 3 bytes of the mesh id, clears them, and
+passes them on to ieee80211_link_info_change_notify, because
+ifmsh->mbss_changed was not updated to match its size.
+Fix this by turning into ifmsh->mbss_changed into an unsigned long array with
+64 bit size.
+
+Fixes: 15ddba5f4311 ("wifi: mac80211: consistently use u64 for BSS changes")
```
Comment (hauke) [inline package/kernel/mac80211/patches/subsys/317-mac80211-fix-mesh-id-corruption-on-32-bit-systems.patch:15]: I think we do not ship `wifi: mac80211: consistently use u64 for BSS changes`. Is this patch really needed?

File: package/kernel/mac80211/patches/subsys/351-mac80211-fix-sband-iftype-data-lookup-for-AP_VLAN.patch [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+---
+
+--- a/include/net/cfg80211.h
++++ b/include/net/cfg80211.h
+@@ -567,6 +567,9 @@ ieee80211_get_sband_iftype_data(const st
+ 	if (WARN_ON(iftype >= NL80211_IFTYPE_MAX))
+ 		return NULL;
+ 
++	if (iftype == NL80211_IFTYPE_AP_VLAN)
++		iftype = NL80211_IFTYPE_AP;
```
Comment (hauke) [inline package/kernel/mac80211/patches/subsys/351-mac80211-fix-sband-iftype-data-lookup-for-AP_VLAN.patch:19]: This patch is already included upstream: https://github.com/gregkh/linux/commit/521860ddf31414a58f2dcd41f04b790603db8c8b

**Comment by rmilecki** [comment]
@dangowrt a kind request to look at those backports

**Comment by rany2** [comment]
Did you intentionally drop 8ef4d785829241e3a6f95c896de210fd5d3f073b ?

**Comment by dangowrt** [comment]
Support for features mostly needed on MT7988 doesn't make much sense for 23.05, as we anyway only support the reference board. Hence please drop patches adding support for multiple PPE, it's a too big change for a stable release to happen on short notice and without much testing imho.

Patches which make sense backporting are:
 * net: ethernet: mtk_eth_soc: handle dma buffer size soc specific
 * net: ethernet: mtk_eth_soc: implement .{get,set}_pauseparam ethtool ops
 * net: ethernet: mtk_ppe: Change PPE entries number to 16K

**Comment by lukasz1992** [comment]
@dangowrt updated; @rany2 In that case dropping it

**Comment by rmilecki** [comment]
Thanks you @dangowrt & @lukasz1992, I'm looking at this right now

**Comment by rany2** [comment]
Nitpicking but all kernel-related commits except [generic: mtk_eth_soc: add support for flow-control settings](https://github.com/openwrt/openwrt/pull/15836/commits/7998fc562b26a303b858c155219d3e8a4c31c373) (7998fc562b26a303b858c155219d3e8a4c31c373) should be in `backport-5.15` not `patches-5.15`.

Also [generic: mtk_eth_soc: add support for flow-control settings](https://github.com/openwrt/openwrt/pull/15836/commits/7998fc562b26a303b858c155219d3e8a4c31c373) (7998fc562b26a303b858c155219d3e8a4c31c373) should come before [generic: mtk_eth_soc: import accepted patches](https://github.com/openwrt/openwrt/pull/15836/commits/8462a71d194b86fc10817f40d8ec406550022699) (8462a71d194b86fc10817f40d8ec406550022699) so moving to `backport-5.15` works.

File: package/kernel/mac80211/patches/subsys/317-mac80211-fix-mesh-id-corruption-on-32-bit-systems.patch (hunk 4 above)
Comment (hauke) [reply package/kernel/mac80211/patches/subsys/317-mac80211-fix-mesh-id-corruption-on-32-bit-systems.patch:15]: Thanks for the link. We should backport this patch.

**Comment by rmilecki** [comment]
I backported `net: ethernet: mtk_eth_soc: handle dma buffer size soc specific` on my own, hopefully the right way. It was terribly annoying as patches were getting mangled.

**Comment by rmilecki** [comment]
Plus the other two `mtk_eth_soc` patches now.

**Comment by lukasz1992** [comment]
So dropping the rest of patches.

**Pull request merged**